### PR TITLE
feat: Add RecordingRule and RecordingRulesUpdate

### PR DESCRIPTION
**Pull request by FMV1491**
# Fixes #

A short description of what this PR does.

### Checklist
- [x] I acknowledge that all my contributions will be made under the project's license
- [ ] I have made a material change to the repo (functionality, testing, spelling, grammar)
- [ ] I have read the [Contribution Guidelines](CONTRIBUTING.md) and my PR follows them
- [ ] I have titled the PR appropriately
- [ ] I have updated my branch with the main branch
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation about the functionality in the appropriate .md file
- [ ] I have added inline documentation to the code I modified

If you have questions, please file a [support ticket](https://twilio.com/help/contact), or create a GitHub Issue in this repository.

### CODE REVIEW

File: src/main/java/com/twilio/type/RecordingRule.java [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+import com.fasterxml.jackson.annotation.JsonValue;
+import com.google.common.base.MoreObjects;
+import com.twilio.converter.Promoter;
+
+import java.util.Objects;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class RecordingRule {
+
+    public enum Type {
```
[the remark this answers is not in the document]
Comment (thinkingserious) [reply src/main/java/com/twilio/type/RecordingRule.java:15]: There are other duplications with the `SubscribeRule` class that might make sense to refactor too.

File: src/main/java/com/twilio/type/RecordingRulesUpdate.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
+package com.twilio.type;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.MoreObjects;
+
+import java.util.List;
+import java.util.Objects;
+
```
Comment (thinkingserious) [inline src/main/java/com/twilio/type/RecordingRulesUpdate.java:9]: Please add a short description here as noted in the `RecordingRule` class.

File: src/main/java/com/twilio/type/RecordingRulesUpdate.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
+package com.twilio.type;
```
Comment (thinkingserious) [inline src/main/java/com/twilio/type/RecordingRulesUpdate.java:1]: There are many similarities to the `SubscribeRulesUpdate.java` class. Similar to the `SubscribeRule` and `RecordingRule` classes, we can simplify with a base class.

**Review by thinkingserious — CHANGES_REQUESTED** [review]
Thank you for the PR!

**Review by innerverse — APPROVED** [review]

File: src/main/java/com/twilio/type/RecordingRulesUpdate.java (hunk 3 above)
Comment (FMV1491) [reply src/main/java/com/twilio/type/RecordingRulesUpdate.java:1]: Done

**Review by innerverse — APPROVED** [review]

File: src/main/java/com/twilio/type/RecordingRule.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        BuilderMiddleBuild withTrack(final String track);
+        BuilderBuild withAll();
+    }
+
+    public interface BuilderMiddleBuild {
+        BuilderMiddleBuild withPublisher(final String publisher);
+        BuilderMiddleBuild withKind(final Kind kind);
+        BuilderMiddleBuild withTrack(final String track);
+        RecordingRule build();
+    }
```
Comment (childish-sambino) [inline src/main/java/com/twilio/type/RecordingRule.java:79]: I don't understand the need for these additional interfaces. And one is returning the type of the other.

Comment (childish-sambino) [reply src/main/java/com/twilio/type/RecordingRule.java:79]: Ah, I see the progression now.

**Review by childish-sambino — APPROVED** [review]

**Pull request merged**